### PR TITLE
GH-26255: Refactor CacheInvalidate sendPurgeRequest to fix incorrect tag splitting

### LINKED ISSUE #8815: Varnish "Connection reset by peer" error when large catalog is reindexed on schedule
### Addition information for Re-Open

The Issue was reopened based on several complaints that it **was not properly fixed** and it still can be reproduced on the `2.2.8` and `2.3.1`
For more details and explanation see next comments:
- https://github.com/magento/magento2/issues/8815#issuecomment-493552366
- https://github.com/magento/magento2/issues/8815#issuecomment-504731799

### Preconditions
1. Magento EE 2.1.5 with ~10K sample products
2. Server is running PHP version 7.0.15
3. All indexes are configured to "UPDATE BY SCHEDULE"
4. Cron job is running on server. Cron error email address is configured with your email so you get notifications of errors.
5. Server is configured to use Varnish for Full Page Cache
6. Varnish is configured with `http_req_hdr_len` and `http_req_size` configured at their default values (per [the documentation](https://www.varnish-cache.org/docs/4.0/reference/varnishd.html))

### Steps to reproduce
1. In the admin, go to _SYSTEM > Export_ and export a CSV of all ~10K products
2. Import the CSV back into Magento using the _SYSTEM > Import_ interface

### Expected result
The import will be processed without any errors.

### Actual result
The cron will have an error when reindexing and will send an error like this to the cron email address:
```
PHP Notice: fwrite(): send of 8192 bytes failed with errno=104 Connection reset by peer in /var/www/stage/releases/20170213214641/vendor/zendframework/zend-http/src/Client/Adapter/Socket.php on line 376
```
The error above is triggered by line 375 of `vendor/zendframework/zend-http/src/Client/Adapter/Socket.php`

In order to determine what request was causing the error, I edited the `Socket.php` file to add logging code at line 375 to log large requests, along with a backtrace:
```
if (strlen($request) > 8000) {
    file_put_contents('/var/www/stage/current/var/log/connection_error.log', date('c') . PHP_EOL . print_r(debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 100), true), FILE_APPEND);
    file_put_contents('/var/www/stage/current/var/log/connection_error.log', date('c') . PHP_EOL . print_r($request, true), FILE_APPEND);
}
```

The requests that were causing errors were ~1.1MB in size, which exceeded the `http_req_size` default Varnish setting (which is 32k). While we could crank up that setting to allow the massive requests to be sent to Varnish, this would increase the Varnish memory footprint and is not an acceptable solution. Read this for context on the memory implications: https://www.varnish-cache.org/docs/4.0/reference/varnishd.html#http-req-size

Here is a preview of what the 1.1MB request contained: 
```
PURGE / HTTP/1.1
X-Magento-Tags-Pattern: ((^|,)catalog_product_6574(,|$))|((^|,)catalog_product_6575(,|$))|((^|,)catalog_product_6576(,|$))|((^|,)catalog_product_6577(,|$))|((^|,)catalog_product_6578(,|$))|((^|,)catalog_product_6579(,|$))|((^|,)catalog_product_6580(,|$))|((^|,)catalog_product_6581(,|$))|((^|,)catalog_product_6582(,|$))|((^|,)catalog_product_6583(,|$))|((^|,)catalog_product_6584(,|$))|((^|,)catalog_product_6585(,|$))|((^|,)catalog_product_6586(,|$))|((^|,)catalog_product_6587(,|$))|((^|,)catalog_product_6588(,|$))|((^|,)catalog_product_6589(,|$))|((^|,)catalog_product_6590(,|$))|((^|,)catalog_product_6591(,|$))|((^|,)catalog_product_6593(,|$))|((^|,)catalog_product_6594(,|$))|((^|,)catalog_product_6595(,|$))|((^|,)catalog_product_6596(,|$))|((^|,)catalog_product_6597(,|$))|((^|,)catalog_product_6598(,|$))|((^|,)catalog_product_6599(,|$))|((^|,)catalog_product_6600(,|$))|((^|,)catalog_product_6601(,|$))|((^|,)catalog_product_6602(,|$))|((^|,)catalog_product_6603(,|$))|((^|,)catalog_product_6604(,|$))|((^|,)catalog_product_6605(,|$))|((^|,)catalog_product_6606(,|$))|((^|,)catalog_product_6607(,|$))|((^|,)catalog_product_6608(,|$))|((^|,)catalog_product_6644(,|$))|((^|,)catalog_product_6645(,|$))|((^|,)catalog_product_6646(,|$))|((^|,)catalog_product_6647(,|$))|((^|,)catalog_product_6648(,|$))|((^|,)catalog_product_6649(,|$))|((^|,)catalog_product_6650(,|$))|((^|,)catalog_product_6651(,|$))|((^|,)
```

In case it is useful for troubleshooting this issue, here is the backtrace logged by the logging code I added to the `Socket.php` file:
```
2017-02-26T19:15:46+00:00
Array
(
    [0] => Array
        (
            [file] => /var/www/stage/releases/20170224215342/vendor/magento/module-cache-invalidate/Model/PurgeCache.php
            [line] => 66
            [function] => write
            [class] => Zend\Http\Client\Adapter\Socket
            [type] => ->
        )

    [1] => Array
        (
            [file] => /var/www/stage/releases/20170224215342/vendor/magento/module-cache-invalidate/Observer/InvalidateVarnishObserver.php
            [line] => 65
            [function] => sendPurgeRequest
            [class] => Magento\CacheInvalidate\Model\PurgeCache
            [type] => ->
        )

    [2] => Array
        (
            [file] => /var/www/stage/releases/20170224215342/vendor/magento/framework/Event/Invoker/InvokerDefault.php
            [line] => 73
            [function] => execute
            [class] => Magento\CacheInvalidate\Observer\InvalidateVarnishObserver
            [type] => ->
        )

    [3] => Array
        (
            [file] => /var/www/stage/releases/20170224215342/vendor/magento/framework/Event/Invoker/InvokerDefault.php
            [line] => 61
            [function] => _callObserverMethod
            [class] => Magento\Framework\Event\Invoker\InvokerDefault
            [type] => ->
        )

    [4] => Array
        (
            [file] => /var/www/stage/releases/20170224215342/vendor/magento/framework/Event/Manager.php
            [line] => 66
            [function] => dispatch
            [class] => Magento\Framework\Event\Invoker\InvokerDefault
            [type] => ->
        )

    [5] => Array
        (
            [file] => /var/www/stage/releases/20170224215342/vendor/magento/module-indexer/Model/Processor/CleanCache.php
            [line] => 48
            [function] => dispatch
            [class] => Magento\Framework\Event\Manager
            [type] => ->
        )

    [6] => Array
        (
            [file] => /var/www/stage/releases/20170224215342/vendor/magento/framework/Interception/Interceptor.php
            [line] => 152
            [function] => afterUpdateMview
            [class] => Magento\Indexer\Model\Processor\CleanCache
            [type] => ->
        )

    [7] => Array
        (
            [file] => /var/www/stage/releases/20170224215342/var/generation/Magento/Indexer/Model/Processor/Interceptor.php
            [line] => 39
            [function] => ___callPlugins
            [class] => Magento\Indexer\Model\Processor\Interceptor
            [type] => ->
        )

    [8] => Array
        (
            [file] => /var/www/stage/releases/20170224215342/vendor/magento/module-indexer/Cron/UpdateMview.php
            [line] => 31
            [function] => updateMview
            [class] => Magento\Indexer\Model\Processor\Interceptor
            [type] => ->
        )

    [9] => Array
        (
            [function] => execute
            [class] => Magento\Indexer\Cron\UpdateMview
            [type] => ->
        )

    [10] => Array
        (
            [file] => /var/www/stage/releases/20170224215342/vendor/magento/module-cron/Observer/ProcessCronQueueObserver.php
            [line] => 247
            [function] => call_user_func_array
        )

    [11] => Array
        (
            [file] => /var/www/stage/releases/20170224215342/vendor/magento/module-cron/Observer/ProcessCronQueueObserver.php
            [line] => 195
            [function] => _runJob
            [class] => Magento\Cron\Observer\ProcessCronQueueObserver
            [type] => ->
        )

    [12] => Array
        (
            [file] => /var/www/stage/releases/20170224215342/vendor/magento/framework/Event/Invoker/InvokerDefault.php
            [line] => 73
            [function] => execute
            [class] => Magento\Cron\Observer\ProcessCronQueueObserver
            [type] => ->
        )

    [13] => Array
        (
            [file] => /var/www/stage/releases/20170224215342/vendor/magento/framework/Event/Invoker/InvokerDefault.php
            [line] => 61
            [function] => _callObserverMethod
            [class] => Magento\Framework\Event\Invoker\InvokerDefault
            [type] => ->
        )

    [14] => Array
        (
            [file] => /var/www/stage/releases/20170224215342/vendor/magento/module-staging/Model/Event/Manager.php
            [line] => 97
            [function] => dispatch
            [class] => Magento\Framework\Event\Invoker\InvokerDefault
            [type] => ->
        )

    [15] => Array
        (
            [file] => /var/www/stage/releases/20170224215342/var/generation/Magento/Staging/Model/Event/Manager/Proxy.php
            [line] => 95
            [function] => dispatch
            [class] => Magento\Staging\Model\Event\Manager
            [type] => ->
        )

    [16] => Array
        (
            [file] => /var/www/stage/releases/20170224215342/vendor/magento/framework/App/Cron.php
            [line] => 74
            [function] => dispatch
            [class] => Magento\Staging\Model\Event\Manager\Proxy
            [type] => ->
        )

    [17] => Array
        (
            [file] => /var/www/stage/releases/20170224215342/vendor/magento/module-cron/Console/Command/CronCommand.php
            [line] => 98
            [function] => launch
            [class] => Magento\Framework\App\Cron
            [type] => ->
        )

    [18] => Array
        (
            [file] => /var/www/stage/releases/20170224215342/vendor/symfony/console/Symfony/Component/Console/Command/Command.php
            [line] => 257
            [function] => execute
            [class] => Magento\Cron\Console\Command\CronCommand
            [type] => ->
        )

    [19] => Array
        (
            [file] => /var/www/stage/releases/20170224215342/vendor/symfony/console/Symfony/Component/Console/Application.php
            [line] => 874
            [function] => run
            [class] => Symfony\Component\Console\Command\Command
            [type] => ->
        )

    [20] => Array
        (
            [file] => /var/www/stage/releases/20170224215342/vendor/symfony/console/Symfony/Component/Console/Application.php
            [line] => 195
            [function] => doRunCommand
            [class] => Symfony\Component\Console\Application
            [type] => ->
        )

    [21] => Array
        (
            [file] => /var/www/stage/releases/20170224215342/vendor/magento/framework/Console/Cli.php
            [line] => 96
            [function] => doRun
            [class] => Symfony\Component\Console\Application
            [type] => ->
        )

    [22] => Array
        (
            [file] => /var/www/stage/releases/20170224215342/vendor/symfony/console/Symfony/Component/Console/Application.php
            [line] => 126
            [function] => doRun
            [class] => Magento\Framework\Console\Cli
            [type] => ->
        )

    [23] => Array
        (
            [file] => /var/www/stage/releases/20170224215342/bin/magento
            [line] => 23
            [function] => run
            [class] => Symfony\Component\Console\Application
            [type] => ->
        )

)
```

### LINKED ISSUE #26255: Magento_CacheInvalidate mis-handles very large tag patterns when doing a PURGE
**Preconditions:**
\Magento\CacheInvalidate\Model\PurgeCache::sendPurgeRequest is responsible for taking arbitrarily large tag patterns and chunking it into HTTP PURGE requests based on the maximum header length Varnish will accept. 

Currently, \Magento\CacheInvalidate\Observer\InvalidateVarnishObserver::execute will join tags with '|' and call sendPurgeRequest. Those tags themselves may include '|' characters.

\Magento\CacheInvalidate\Model\PurgeCache::splitTags will then separate on '|' and re-join components until the length of those components reaches the max. header length. This is too naive, since it may now split an individual tag (one which itself included '|' characters) across two requests. Both of those requests now have an invalid pattern - this would mean that at the very least, some purges will be missed. In the worst-case scenario, I expect this could also lead to Varnish being completely purged, depending on the tag that was incorrectly split.

For the sake of brevity, the example below has been limited to a small subset of my real-world example; as such let's pretend the max header size is 256 bytes. Note that this is a typical pattern that would result from a product save event.

`((^|,)cat_p_1468(,|$))|((^|,)cat_p_1361(,|$))|((^|,)cat_p_1473(,|$))|((^|,)cat_p_930(,|$))|((^|,)cat_p_933(,|$))|((^|,)cat_p_934(,|$))|((^|,)cat_p_664(,|$))|((^|,)cat_p_487(,|$))|((^|,)cat_p_490(,|$))|((^|,)cat_p_491(,|$))|((^|,)cat_p_153(,|$))|((^|,)cat_p_156(,|$))|((^|,)cat_p_157(,|$))|((^|,)cat_p_497(,|$))|((^|,)cat_p_498(,|$))|((^|,)cat_p_499(,|$))|((^|,)cat_p_227(,|$))|((^|,)cat_p_228(,|$))|((^|,)cat_p_229(,|$))|((^|,)cat_p_242(,|$))|((^|,)cat_p_244(,|$))|((^|,)cat_p_245(,|$))`

Based on my desired max. header size, this will be chunked into two separate requests:

`((^|,)cat_p_1468(,|$))|((^|,)cat_p_1361(,|$))|((^|,)cat_p_1473(,|$))|((^|,)cat_p_930(,|$))|((^|,)cat_p_933(,|$))|((^|,)cat_p_934(,|$))|((^|,)cat_p_664(,|$))|((^|,)cat_p_487(,|$))|((^|,)cat_p_490(,|$))|((^|,)cat_p_491(,|$))|((^|,)cat_p_153(,|$))|((^`

`,)cat_p_156(,|$))|((^|,)cat_p_157(,|$))|((^|,)cat_p_497(,|$))|((^|,)cat_p_498(,|$))|((^|,)cat_p_499(,|$))|((^|,)cat_p_227(,|$))|((^|,)cat_p_228(,|$))|((^|,)cat_p_229(,|$))|((^|,)cat_p_242(,|$))|((^|,)cat_p_244(,|$))|((^|,)cat_p_245(,|$))|((^|,)cat_p_245(,|$))`

Expected split behaviour:
`((^|,)cat_p_1468(,|$))|((^|,)cat_p_1361(,|$))|((^|,)cat_p_1473(,|$))|((^|,)cat_p_930(,|$))|((^|,)cat_p_933(,|$))|((^|,)cat_p_934(,|$))|((^|,)cat_p_664(,|$))|((^|,)cat_p_487(,|$))|((^|,)cat_p_490(,|$))|((^|,)cat_p_491(,|$))|((^|,)cat_p_153(,|$))`

`((^|,)cat_p_156(,|$))|((^|,)cat_p_157(,|$))|((^|,)cat_p_497(,|$))|((^|,)cat_p_498(,|$))|((^|,)cat_p_499(,|$))|((^|,)cat_p_227(,|$))|((^|,)cat_p_228(,|$))|((^|,)cat_p_229(,|$))|((^|,)cat_p_242(,|$))|((^|,)cat_p_244(,|$))|((^|,)cat_p_245(,|$))`

---
**Steps to reproduce:**
- Have full_page cache enabled
- Have caching application configured to Varnish
- Run code below
```
<?php

require __DIR__ . '/app/bootstrap.php';
$bootstrap = \Magento\Framework\App\Bootstrap::create(BP, $_SERVER);

class DummyEntity implements \Magento\Framework\DataObject\IdentityInterface
{
    public function getIdentities()
    {
        $ids = range(807, 1193);
        return array_map(function ($id) {
            return 'cat_p_' . $id;
        }, $ids);
    }
}

$objectManager = $bootstrap->getObjectManager();
$appState = $objectManager->get('Magento\Framework\App\State');
$appState->setAreaCode('frontend');
$eventManager = $objectManager->get('\Magento\Framework\Event\Manager');

$myObject = new DummyEntity();

$eventManager->dispatch('clean_cache_by_tags', ['object' => $myObject]);
```
**Actual Result:**
Observe two resulting PURGE requests sent to Varnish with invalid patterns (e.g. xdebug_break inside \Magento\CacheInvalidate\Model\PurgeCache::sendPurgeRequestToServers and observe the X-Magento-Tags-Pattern header)

**Expected Result:**

### PR DESCRIPTION
**Pull request by moloughlin**
### Description (*)
Issue: https://github.com/magento/magento2/issues/26255

- Refactored sendPurgeRequest to take in an array of tags instead of requiring the caller to implode(). This allows us to do smarter chunking, and avoids the incorrect splitting noted in the linked issue.
- Backwards compatibility maintained for callers passing in a string, calls from core code have been updated to use the new array form
- Allowed header size to be passed in via DI

### Fixed Issues (if relevant)
1. Fixes magento/magento2#26255
2. Fixes magento/magento2#8815 : Varnish "Connection reset by peer" error when large catalog is reindexed on schedule 

### Manual testing scenarios (*)
1. See https://github.com/magento/magento2/issues/26255

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/Magento/CacheInvalidate/Model/PurgeCache.php [hunk 1]
```diff
@@ -39,39 +39,46 @@ class PurgeCache
...
      * @param \Magento\CacheInvalidate\Model\SocketFactory $socketAdapterFactory
      * @param InvalidateLogger $logger
+     * @param int $maxHeaderSize
      */
     public function __construct(
         \Magento\PageCache\Model\Cache\Server $cacheServer,
         \Magento\CacheInvalidate\Model\SocketFactory $socketAdapterFactory,
-        InvalidateLogger $logger
+        InvalidateLogger $logger,
+        $maxHeaderSize = 7680
```
Comment (lbajsarowicz) [inline app/code/Magento/CacheInvalidate/Model/PurgeCache.php]: `int` ?

File: app/code/Magento/CacheInvalidate/Model/PurgeCache.php [hunk 2]
```diff
@@ -39,39 +39,46 @@ class PurgeCache
...
      * Send curl purge request to invalidate cache by tags pattern
      *
-     * @param string $tagsPattern
+     * @param array|string $tags
      * @return bool Return true if successful; otherwise return false
      */
-    public function sendPurgeRequest($tagsPattern)
+    public function sendPurgeRequest($tags)
     {
+        if (!is_string($tags)) {
```
Comment (lbajsarowicz) [inline app/code/Magento/CacheInvalidate/Model/PurgeCache.php]: Why `!`? That's not gonna work properly.

**Review by lbajsarowicz — DISMISSED** [review]
Could you provide some Unit Tests for that changes?

File: app/code/Magento/CacheInvalidate/Model/PurgeCache.php (hunk 2 above)
Comment (moloughlin) [reply app/code/Magento/CacheInvalidate/Model/PurgeCache.php]: Yep, my head wasn't screwed on right when I committed that :smile: it has since been fixed

**Comment by moloughlin** [comment]
@lbajsarowicz, I've added a test for the splitting of large tag arrays and fixed the other outstanding suggestions. Ready for another look when you are available - thanks!

**Comment by moloughlin** [comment]
@lbajsarowicz can you please confirm next steps on this? Thanks.

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
@moloughlin Let's wait for tests to pass.
Code looks fine, so I'll approve the PR after passing the tests.

Could you eventually merge latest `2.4-develop` to your working branch?

**Comment by moloughlin** [comment]
@lbajsarowicz awesome, thanks. 2.4-develop merged back in and only the typical false negatives on the test runs :+1:

**Comment by moloughlin** [comment]
Bump (merged 2.4-develop again since it's been a few weeks...)

**Comment by lbajsarowicz** [comment]
@moloughlin Thanks for your contribution (again).
We faced false-negative with your PR Static Tests.

After tests pass - I'm going to approve the PR.

**Review by lbajsarowicz — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-7073](https://jira.corp.magento.com/browse/ENGCOM-7073) has been created to process this Pull Request
:eight_spoked_asterisk: @lbajsarowicz, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

File: app/code/Magento/CacheInvalidate/Model/PurgeCache.php [hunk 3]
```diff
@@ -39,39 +39,46 @@ class PurgeCache
...
         \Magento\PageCache\Model\Cache\Server $cacheServer,
         \Magento\CacheInvalidate\Model\SocketFactory $socketAdapterFactory,
-        InvalidateLogger $logger
+        InvalidateLogger $logger,
+        int $maxHeaderSize = 7680
     ) {
         $this->cacheServer = $cacheServer;
         $this->socketAdapterFactory = $socketAdapterFactory;
         $this->logger = $logger;
+        $this->requestSize = $maxHeaderSize;
```
Comment (ihor-sviziev) [inline app/code/Magento/CacheInvalidate/Model/PurgeCache.php]: I think it's better to have single name for one thing. Can we name it as `requestSize` or `maxHeaderSize` in all places?

File: app/code/Magento/CacheInvalidate/Model/PurgeCache.php [hunk 4]
```diff
@@ -82,24 +89,24 @@ public function sendPurgeRequest($tagsPattern)
...
     /**
-     * Split tags by batches
+     * Split tags into batches to suit Varnish max. header size
      *
-     * @param string $tagsPattern
+     * @param array $tags
      * @return \Generator
      */
-    private function splitTags($tagsPattern)
+    private function chunkTags($tags)
```
Comment (ihor-sviziev) [inline app/code/Magento/CacheInvalidate/Model/PurgeCache.php]: Looks like we can define argument and return types here:
```suggestion
    private function chunkTags(array $tags): \Generator
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @moloughlin,
Your changes looks really good to me!
Could you review my comments and update your PR?

File: app/code/Magento/CacheInvalidate/Model/PurgeCache.php (hunk 4 above)
Comment (moloughlin) [reply app/code/Magento/CacheInvalidate/Model/PurgeCache.php]: Done (and also on the other private method) - thanks!

**Comment by bradleybrecher** [comment]
@magento give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @bradleybrecher. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Review by ihor-sviziev — APPROVED** [review]
Failing tests not related to changes from this PR

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7073](https://jira.corp.magento.com/browse/ENGCOM-7073) has been created to process this Pull Request

**Comment by engcom-Kilo** [comment]
@moloughlin 
Please resolve conflict with files
app/code/Magento/CacheInvalidate/Test/Unit/Model/PurgeCacheTest.php
app/code/Magento/CacheInvalidate/Test/Unit/Observer/FlushAllCacheObserverTest.php

**Comment by moloughlin** [comment]
> Please resolve conflict with files

@engcom-Kilo Resolved & merged, thanks.

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7073](https://jira.corp.magento.com/browse/ENGCOM-7073) has been created to process this Pull Request

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7073](https://jira.corp.magento.com/browse/ENGCOM-7073) has been created to process this Pull Request

**Comment by ihor-sviziev** [comment]
@engcom-Kilo could you test this PR?

**Comment by engcom-Kilo** [comment]
> @engcom-Kilo could you test this PR?

Checking this.

**Comment by engcom-Kilo** [comment]
✔️ QA Passed

Changes implemented in Pull Request works correctly.


console

**Comment by engcom-Kilo** [comment]
@magento run all tests

**Comment by engcom-Kilo** [comment]
Failed Functional Tests B2B is not related to changes in current Pull Request.

**Review by VladimirZaets — CHANGES_REQUESTED** [review]
Hi, @engcom-Kilo are tested this fix with varnish? If not please retest it.

**Comment by engcom-Kilo** [comment]
@VladimirZaets 
Yes, this was tested with Varnish. Works correctly.

**Review by ihor-sviziev — APPROVED** [review]
Approved again in order to have an ability to move back to accept

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7073](https://jira.corp.magento.com/browse/ENGCOM-7073) has been created to process this Pull Request

**Pull request merged**